docs/proposals/002-refresh-certificates.md

```diff
@@ -0,0 +1,371 @@
+cluster.
+
+### `POST /refresh-certs/run`
+
+```go
+type RefreshCertificatesRunRequest struct {
+  // Seed must match the seed from the received RefreshCertificatesInitResponse.
+  Seed int
+  // ExpirationDuration is the duration of the requested certificates.
+  ExpirationDuration time.Duration
+}
+
+type RefreshCertificatesRunResponse struct {
+  // ExpirationDuration is the duration of the new certificates.
+  ExpirationDuration time.Duration
+}
+```
+
+#### Worker nodes
+
+Worker nodes will create the required CSRs, and wait for them to be signed. This
+comprises of two steps: first they must be approved (see below), and then the
+csrsigning controller will sign them. The worker node will block until the CSRs
+are dealt with:
+
+- If any CSR becomes Failed or Denied, then we fail and return the result.
+- If any CSR does not become Approved or Issued in time, we time out.
+
+After getting the new certificates, we apply them on the node, do any service
+restarts that may be required and return response.
+
+#### Control plane nodes
+
+Control plane nodes can ignore the seed argument. They already have access to
+the CA certificates and keys, therefore can simply generate and sign new
+certificates, apply them on the node, do any service restarts necessary and
+return the response.
+
+## CLI Changes
+<!--
+This section MUST mention any changes to the k8s CLI, e.g. new arguments,
+different outputs.
+-->
+
+### `k8s refresh-certs` command
+
+#### For worker nodes
+
+```
+root@worker:~$ k8s refresh-certs --ttl 30d --timeout 30s
+The following CertificateSigningRequests should be approved. Run the following
+commands on any of the control plane nodes of the cluster:
+
+  $ k8s kubectl certificate approve k8sd-seed-worker-kubelet-client
+  $ k8s kubectl certificate approve k8sd-seed-worker-kubelet-serving
+  $ k8s kubectl certificate approve k8sd-seed-worker-kube-proxy-client
+
+Waiting for certificates to be created....
+
+<-- CSRs are approved -->
+
+Certificates have been refreshed, and will expire at $date.
+```
+
+#### For control plane nodes
+
+```
+root@node:~$ k8s refresh-certs --ttl 30d --timeout 30s
+Certificates have been refreshed, and will expire at $date.
+```
+
+## Database Changes
+<!--
+This section MUST mention any changes required in the k8sd database schema or
+internal types.
+-->
+none
+
+## Configuration Changes
+<!--
+This section MUST mention any new configuration options or service arguments
+that are introduced.
+-->
+
+The changes include a `csrsigning` controller, which watches the cluster for
+`CertificateSigningRequest` objects coming from the worker nodes. After they are
+approved, it will sign the certificates.
+
+### k8sd `--disable-csrsigning-controller`
+
+In line with the rest of the `--disable-$foo-controller` flags, this flag can be
+used to completely disable the csrsigning controller on a node.
+
+### annotation `k8sd/v1alpha1/csrsigning/auto-approve=true`
```

Comment:
As discussed, I'd go with a cluster config option here.

Reply:
annotations can help buy us some time before we know all the configs we might need/want or care about. there are more, e.g. `max-ttl`. i'd start with annotations, and we can move them to stable fields in the cluster config as we iterate on this feature

Reply:
I'd say we should go with annotations first especially while we figure things out. Once something like this gets into the actual config people might expect it to be "officially" supported. I believe we are still in an early stage for this, I'd be up to discussing/working more on this to be official after we get the actual certificate refresh process in place.

Reply:
I'm +1 on annotations. I believe this could help us "graduate" new functionalities we add without messing up the internal cluster configuration, following a process similar to the upstream K8s API.

Reply:
democracy - annotations it is